packages/babel-core/src/config/caching-a.js

```diff
@@ -0,0 +1,108 @@
+// @flow
+
+import gensync, { type Handler, type Gensync } from "gensync";
+import { CacheConfigurator } from "./caching";
+
+export type SimpleCacheConfigurator = SimpleCacheConfiguratorFn &
+  SimpleCacheConfiguratorObj;
+
+type SimpleCacheConfiguratorFn = {
+  (boolean): void,
+  <T>(handler: () => T): T,
+};
+type SimpleCacheConfiguratorObj = {
+  forever: () => void,
+  never: () => void,
+  using: <T>(handler: () => T) => T,
+  invalidate: <T>(handler: () => T) => T,
+};
+
+type CacheEntry<ResultT, SideChannel> = Array<{
+  value: ResultT,
+  valid: SideChannel => boolean,
+}>;
+
+export type { CacheConfigurator };
+
+/**
+ * Given a function with a single argument, cache its results based on its argument and how it
+ * configures its caching behavior. Cached values are stored strongly.
+ */
+export function makeStrongCache<ArgT, ResultT, SideChannel>(
+  handler: (ArgT, CacheConfigurator<SideChannel>) => Handler<ResultT>,
+): Gensync<[ArgT, SideChannel], ResultT> {
+  return makeCachedFunction<ArgT, ResultT, SideChannel, *>(new Map(), handler);
+}
+
+/**
+ * Given a function with a single argument, cache its results based on its argument and how it
+ * configures its caching behavior. Cached values are stored weakly and the function argument must be
+ * an object type.
+ */
+export function makeWeakCache<
+  ArgT: {} | Array<*> | $ReadOnlyArray<*>,
+  ResultT,
+  SideChannel,
+>(
+  handler: (ArgT, CacheConfigurator<SideChannel>) => Handler<ResultT>,
+): Gensync<[ArgT, SideChannel], ResultT> {
+  return makeCachedFunction<ArgT, ResultT, SideChannel, *>(
+    new WeakMap(),
+    handler,
+  );
+}
+
+type CacheMap<ArgT, ResultT, SideChannel> =
+  | Map<ArgT, CacheEntry<ResultT, SideChannel>>
+  | WeakMap<ArgT, CacheEntry<ResultT, SideChannel>>;
+
+function makeCachedFunction<
+  ArgT,
+  ResultT,
+  SideChannel,
+  // $FlowIssue https://github.com/facebook/flow/issues/4528
+  Cache: CacheMap<ArgT, ResultT, SideChannel>,
+>(
+  callCache: Cache,
+  handler: (ArgT, CacheConfigurator<SideChannel>) => Handler<ResultT>,
+): Gensync<[ArgT, SideChannel], ResultT> {
+  return gensync(function* cachedFunction(arg: ArgT, data: SideChannel) {
+    let cachedValue: CacheEntry<ResultT, SideChannel> | void = callCache.get(
+      arg,
+    );
+
+    if (cachedValue) {
+      for (const { value, valid } of cachedValue) {
+        if (valid(data)) return value;
+      }
+    }
+
+    const cache = new CacheConfigurator(data);
+
+    const value = yield* handler(arg, cache);
```

Comment:
One big complexity here is that, if this is async, you can end up with multiple calls to `handler` running in parallel. In my opinion, we should only ever allow one call to `handler` at a time for a given `arg`.

Reply:
I'm not sure about how they could run in parallel. If it is async, `handler`'s effects will be yielded and no other code will be run until they are completed, including other `handler` calls.

Reply:
For instance, if you do
```
var doThing = makeWeakCache(function*(arg) {
  yield* delay(100);
  return {};
});

var obj = {};
const [r1, r2] = yield* gensync.all[doThing(obj), doThing(obj)]);

console.log(r1 === r2);
```
would be `true` in sync code and `false` in async code with your current implementation because the first would run then populate the cache with the `r1` object, but then the second one would finish and replace the `callCache` data with `r2`. So really the second call to see that there is already a call in-progress and wait for it to finish before checking if the cache entry matches.

Reply:
Was this explanation enough to make my concern clearer?

Reply:
Yes sorry, I just didn't have time to update it! There is a reason if this PR is marked as "draft" 😛

Reply:
# TODO

Reply:
@loganfsmyth I was implementing this, and I'm not sure that it is something that we want, since it can cause a few problems:
1. If a function is called recursively (like a plugin loading another plugin via `inherits`), it will cause a deadlock: the outer function needs the inner one's return value, but the inner one won't run until the outer one is finished.
2. It will make async calls to Babel slower. Suppose that we have this code:
    ```js
    var maybeDelay = gensync({
      sync: () => {},
      async: t => delay(t),
    });

    var doThing = gensync(makeWeakCache(function*(arg) {
      yield* maybeDelay(100);
      return {};
    }));

	doThing.async();
    await delay(50);
    doThins.sync();
    ```

    here `doThins.sync()` runs while `doThing.async` is paused by the `delay` call. If we can't call `doThing.sync` until `doThing.async` is finished, the only option we have is to throw an error, because it can't wait.

    `maybeDelay` might seem like a function crafted only to create this example, without real-world applications, but it is exactly like any `fs.*` operation.

I think that randomly throwing errors caused by race conditions is a too hight cost for to be introduced only for performance reasons. I think that we have two options:
1. Leave is as-is, allowing two different async calls to happen in parallel
2. Always invalidate the cache when switching between sync/async. I already had to do it when interacting with an either sync or async `cache.invalidate` call, but it shouldn't be observable.

Reply:
The big issue for me is that I could really see this being a noticable perf hit for something that starts compiling a bunch of files in parallel, like `@babel/cli`. If you do something like

```
var results = Promise.all(files.map(filename => babel.transformFileAsync(filename)));
```
then you'll likely end up calling every plugin and preset function for every single file, because config resolving will all be kicked off in parallel and each file's process will kick off an operation that then won't get the cached valid of the previous file compilation get, because the operation will have started before the previous one finished. I haven't actually tested that though, so if you want to verify that could help the discussion. Maybe I'm missing something.

> If a function is called recursively (like a plugin loading another plugin via inherits), it will cause a deadlock: the outer function needs the inner one's return value, but the inner one won't run until the outer one is finished.

We should probably disallow that anyway, if we don't already, since it would be an unresolvable problem in sync code too. I agree that recursion could be a concern, but I'm not sure it's one that wouldn't already be an issue in sync cases in general too.

> If we can't call doThing.sync until doThing.async is finished, the only option we have is to throw an error, because it can't wait.

I think it's probable find to have entirely separate caching for sync and async calls to avoid this conflict. In fact we likely explicitly want that, because plugins could return different results depending on whether the system is sync or async.

Reply:
Sounds like you need a mutex.